drivers/serial/uart_stm32.c

```diff
@@ -1501,10 +1516,22 @@ static void uart_stm32_irq_config_func_##index(const struct device *dev)	\
 }
 #else
 #define STM32_UART_IRQ_HANDLER_DECL(index)
-#define STM32_UART_IRQ_HANDLER_FUNC(index)
 #define STM32_UART_IRQ_HANDLER(index)
 #endif
 
+#if defined(CONFIG_UART_INTERRUPT_DRIVEN) || defined(CONFIG_UART_ASYNC_API)
+#define STM32_UART_IRQ_HANDLER_FUNC(index)				\
+	.irq_config_func = uart_stm32_irq_config_func_##index,
+#define STM32_UART_POLL_IRQ_HANDLER_FUNC(index)
+#elif defined(CONFIG_PM)
+#define STM32_UART_IRQ_HANDLER_FUNC(index)
+#define STM32_UART_POLL_IRQ_HANDLER_FUNC(index)				\
+	.irq_config_func = uart_stm32_irq_config_func_##index,
+#else
+#define STM32_UART_IRQ_HANDLER_FUNC(index)
+#define STM32_UART_POLL_IRQ_HANDLER_FUNC(index)
+#endif
```

Comment:
Nitpick (I'm thinking to the ones that will read these lines for the first time)

```suggestion
#if defined(CONFIG_UART_INTERRUPT_DRIVEN) || defined(CONFIG_UART_ASYNC_API)
#define STM32_UART_IRQ_HANDLER_FUNC(index)				\
	.irq_config_func = uart_stm32_irq_config_func_##index,
#define STM32_UART_POLL_IRQ_HANDLER_FUNC(index) /* Not used */
#elif defined(CONFIG_PM)
#define STM32_UART_IRQ_HANDLER_FUNC(index)     /* Not used */
#define STM32_UART_POLL_IRQ_HANDLER_FUNC(index)			\
	.irq_config_func = uart_stm32_irq_config_func_##index,
#else
#define STM32_UART_IRQ_HANDLER_FUNC(index)      /* Not used */
#define STM32_UART_POLL_IRQ_HANDLER_FUNC(index) /* Not used */
#endif
```

Reply:
done